icu4c/source/common/loclikely.cpp

```diff
@@ -483,7 +483,16 @@ _uloc_addLikelySubtags(const char* localeID,
         if(U_FAILURE(*err)) {
             goto error;
         }
-        icu::LSR lsr = likelySubtags->makeMaximizedLsrFrom(icu::Locale::createFromName(localeID), true, *err);
+        // We need to keep l on the stack because lsr may point into internal
+        // memory of l.
```

Comment:
Is declaring `l` as a local variable sufficient to do that?  I don't think you're guaranteed the object will persist until the end of the function if you don't reference it again-- I think the compiler is free to destruct `l` earlier, although I don't know if any of them actually do that.

Reply:
Yes, because `l` will exist until the `}` of the enclosing block, whereas previously it existed only until the `;` on the line it was used as a transient variable.

Reply:
the lsr.language, lsr.region and lsr.script will reference to the memory returned by the getLanguage(), getScript()  and getCountry() of the Locale passed into makeMaximizedLsrFrom. So if we do not keep it as a local on the stack, those value became dangling pointers after the call to makeMaximizedLsrFrom and while we use it in the uprv_strcmp (line 488/497), or uprv_strlen(line 493-506/502-506) it became illegal read. But if we keep it on the stack, they are not illegal read until line 506/515 but we no longer need them at that point.

Reply:
Yeah, you guys are right.  I know that in some languages, the compiler is free to deallocate objects _before_ the closing brace if they know they're not used, but if that were true in C++, all kinds of common C++ idioms would break.  So never mind.  :-)